src/raw.rs

```diff
@@ -27,415 +313,572 @@ where
         }
     }
+}
 
-    pub fn key_only(mut self) -> Self {
-        self.key_only = true;
-        self
+impl RequestInner for ScanInner {
+    type Resp = Vec<KvPair>;
+
+    fn execute(self, client: Arc<RpcClient>, cf: Option<ColumnFamily>) -> KvFuture<Self::Resp> {
+        let keys = match self.range.into_keys() {
+            Err(e) => return Box::new(future::err(e)),
+            Ok(v) => v,
+        };
+        Box::new(client.raw_scan(keys, self.limit, self.key_only, cf))
+    }
+}
+
+/// An unresolved [`Client::scan`](struct.Client.html#method.scan) request.
+///
+/// Once resolved this request will result in a scanner over the given range.
+pub struct Scan {
+    request: Request<ScanInner>,
+}
+
+impl Scan {
+    fn new(client: Arc<RpcClient>, inner: ScanInner) -> Self {
+        Scan {
+            request: Request::new(client, inner),
+        }
     }
 
+    /// Set the (optional) [`ColumnFamily`](struct.ColumnFamily.html).
     pub fn cf(mut self, cf: impl Into<ColumnFamily>) -> Self {
-        self.cf = Some(cf.into());
+        self.request = self.request.cf(cf);
         self
     }
 
-    pub fn reverse(mut self) -> Self {
-        self.reverse = true;
+    pub fn key_only(mut self) -> Self {
+        self.request.inner = self.request.inner.map(|mut x| {
+            x.1.key_only = true;
+            x
+        });
         self
     }
 }
 
-impl<'a> Future for Scan<'a> {
+impl Future for Scan {
     type Item = Vec<KvPair>;
-    type Error = ();
+    type Error = Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let _ = &self.client;
-        let _ = &self.range;
-        let _ = &self.limit;
-        let _ = &self.key_only;
-        let _ = &self.cf;
-        unimplemented!()
+        self.request.poll()
     }
 }
 
-pub struct BatchScan<'a> {
-    client: &'a Client,
-    ranges: Vec<(Key, Key)>,
+pub(crate) struct BatchScanInner {
+    ranges: Vec<Result<(Key, Option<Key>)>>,
     each_limit: u32,
     key_only: bool,
-    cf: Option<ColumnFamily>,
-    reverse: bool,
 }
 
-impl<'a> BatchScan<'a> {
-    fn new(client: &'a Client, ranges: Vec<(Key, Key)>, each_limit: u32) -> Self {
-        BatchScan {
-            client,
+impl BatchScanInner {
+    fn new(ranges: Vec<Result<(Key, Option<Key>)>>, each_limit: u32) -> Self {
+        BatchScanInner {
             ranges,
             each_limit,
             key_only: false,
-            cf: None,
-            reverse: false,
         }
     }
+}
 
-    pub fn key_only(mut self) -> Self {
-        self.key_only = true;
-        self
+impl RequestInner for BatchScanInner {
+    type Resp = Vec<KvPair>;
+
+    fn execute(self, client: Arc<RpcClient>, cf: Option<ColumnFamily>) -> KvFuture<Self::Resp> {
+        let (mut errors, ranges): (Vec<_>, Vec<_>) =
+            self.ranges.into_iter().partition(Result::is_err);
+        if !errors.is_empty() {
+            Box::new(future::err(errors.pop().unwrap().unwrap_err()))
```

Comment:
What about the other errors?

Reply:
They should all be InvalidKeyRange errors, so we can simply pick either one to return. So looks like there is even no need to pop anyone, simply returning a new error would be just good enough and easier.